File: geocoding.R
library(dplyr)
library(readr)
library(rgdal)

# Read in example data and province shape files
df <- read_csv("ACLED.csv") %>% select(event_date, latitude, longitude)
province_shapes <- readOGR("provicinial-boundaries")
provinces <- province_shapes$PRV_NAME %>% as.character()
province_shapes$PRV_NAME <- province_shapes$PRV_NAME %>% as.character()

# Geocode observations by province.
df$province <- NA
for(i in 1:length(provinces)){
  spdf <- df
  coordinates(spdf) <- c("longitude", "latitude")
  proj4string(spdf) <- proj4string(province_shapes[i,])
  idx <- !is.na(over(as(spdf, "SpatialPoints"),
                     as(province_shapes[i,], "SpatialPolygons")))
  df$province[idx] <- province_shapes$PRV_NAME[i]
}
df$province %>% table()

# Filter observations to a specific province
df <- df %>% filter(province %in% c("KUNAR", "NANGARHAR"))

File: geocoding.py
import pandas as pd
import geopandas as gpd

# Read in example data and province shape files
df = pd.read_csv("ACLED.csv")
df = df[['event_date', 'latitude', 'longitude']]
province_shapes = gpd.read_file("provicinial-boundaries")
provinces = province_shapes['PRV_NAME'].tolist()

# Geocode observations by province
gdf = gpd.GeoDataFrame(df, geometry=gpd.points_from_xy(df.longitude, df.latitude))
df = []
for value in range(0,len(province_shapes)):    
    temp = gdf[gdf.geometry.intersects(province_shapes.iloc[value]['geometry'])]
    temp = temp.assign(province=province_shapes.iloc[value]['PRV_NAME'])
    df.append(temp)
df = pd.concat(df)
df['province'].value_counts()

# Filter obseravations to a specific province
df = df[df['province'].isin(['KUNAR', 'NANGARHAR'])]
